### PR TITLE
Category calendar view

### PR DESCRIPTION
**Pull request by mic4ael**

### CODE REVIEW

File: indico/web/assets/bundles.py [hunk 1]
```diff
@@ -237,6 +237,8 @@ def rjs_bundle(name, *files):
                        filters='cssmin', output='css/selectize_css_%(version)s.min.css')
 
 taggle_js = rjs_bundle('taggle_js', 'js/lib/taggle.js')
+fullcalendar_js = rjs_bundle('fullcalendar_js', 'js/lib/fullcalendar.js')
+fullcalendar_css = Bundle('css/')
```
Comment (ThiefMaster) [inline indico/web/assets/bundles.py]: that looks wrong

File: indico/web/assets/bundles.py [hunk 2]
```diff
@@ -506,7 +508,8 @@ def register_all_css(env, main_css_file):
                    'lib/jquery.typeahead.css',
                    'jquery.colorbox.css',
                    'jquery-ui-custom.css',
-                   'jquery.colorpicker.css'),
+                   'jquery.colorpicker.css',
+                   'lib/fullcalendar.css'),
```
Comment (ThiefMaster) [inline indico/web/assets/bundles.py]: why not put it next to the other `lib/` css files?

File: indico/web/assets/bundles.py (hunk 1 above)
Comment (mic4ael) [inline indico/web/assets/bundles.py]: Ah, probably forgot to remove it.

File: indico/modules/categories/templates/display/calendar.html [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+                </a>
+            </span>
+            <div class="category-title" style="margin-left:0px; text-align:left;">
+                {{ category.title }}&nbsp;
+                <span style="font-style: italic; font-size: 0.8em;">({% trans %}Calendar{% endtrans %})</span>
+            </div>
+        </div>
+    </div>
+    <div class="category-content-wrapper category-statistics center" style="text-align: left;">
+        <div style="margin: 0 auto;width: 500px;">
```
Comment (nop33) [inline indico/modules/categories/templates/display/calendar.html]: I guess you just didn't want to bother moving the inline CSS to files, right? That's fine, I just hope you didn't write these lines 😝

Comment (mic4ael) [inline indico/modules/categories/templates/display/calendar.html]: I simply forgot to move those lines to a `scss` file ;)

File: fabfile.py [hunk 4]
```diff
@@ -445,8 +445,20 @@ def install_typewatch():
         local('cp jquery.typewatch.js {}'.format(dest_js_dir))
 
 
-# Tasks
+@recipe('fullcalendar.js')
+def install_fullcalendar_js():
+    with node_env():
```
Comment (ThiefMaster) [inline fabfile.py]: just fyi, you could also do it like this:

    with node_env(), lcd(os.path.join(env.ext_dir, 'fullcalendar'));

File: indico/modules/categories/controllers/display.py [hunk 5]
```diff
@@ -330,3 +332,23 @@ def _process(self):
...
+            category_events = get_category_timetable([self.category.id], form.from_dt.data, form.to_dt.data)['events']
+            tz = self.category.display_tzinfo
+            for day, event_data in category_events.iteritems():
+                for dt, ev in event_data:
+                    if ev.id in serialized_events:
+                        continue
+                    serialized_events[ev.id] = {'title': ev.title, 'start': format_datetime(ev.start_dt, timezone=tz),
+                                                'end': format_datetime(ev.end_dt, timezone=tz),
+                                                'url': url_for('event.conferenceDisplay', ev)}
+        return WPCategoryStatistics.render_template('display/calendar.html',
```
Comment (ThiefMaster) [inline indico/modules/categories/controllers/display.py]: `WPCategory` - this has nothing to do with statistics :)

        return WPCategory.render_template('display/calendar.html', self.category, form=form,
                                          serialized_events=serialized_events.values())

File: indico/modules/categories/forms.py [hunk 6]
```diff
@@ -175,3 +175,12 @@ def validate_entries(self, field):
                 raise ValidationError(_('Invalid category: {}').format(entry['id']))
             if entry['type'] == 'event' and not Event.get(entry['id'], is_deleted=False):
                 raise ValidationError(_('Invalid event: {}').format(entry['id']))
+
+
+class CategoryCalendarForm(IndicoForm):
+    from_dt = IndicoDateTimeField(_('From'))
```
Comment (ThiefMaster) [inline indico/modules/categories/forms.py]: Should be just a date field

**Comment by ThiefMaster** [comment]
Some issues I found:

- Having a start and end date doesn't make much sense. Ideally you would only select a full month. I found this hack on SO to do it with the jQuery date picker, maybe that could be an option. http://jsfiddle.net/DBpJe/5106/
- The calendar should show the selected month by default, and the next/prev buttons should be either disabled or reload the page with the proper month selected
- GET makes more sense than POST here. It'll also make the previous point easier since you can just make the prev/next buttons links (or if they are js-backed redirect using JS to those urls)

**Comment by ThiefMaster** [comment]
Use `Event.is_visible_in(category)` to take visibility into account. You'd do that in `get_category_timetable`.. Check the places where `is_visible_from` is called; instead of that check you need to adapt the queries to properly do the check using the event (might need to join `event_new` for the contribution query)

File: indico/modules/categories/blueprint.py [hunk 7]
```diff
@@ -70,6 +71,7 @@
 _bp.add_url_rule('/events.ics', 'export_ical', RHExportCategoryICAL)
 _bp.add_url_rule('/events.atom', 'export_atom', RHExportCategoryAtom)
 _bp.add_url_rule('/events.rss', 'export_rss', make_compat_redirect_func(_bp, 'export_atom'))
+_bp.add_url_rule('/calendar', 'calendar', RHCategoryCalendarView, methods=('GET', 'POST'))
```
Comment (ThiefMaster) [inline indico/modules/categories/blueprint.py]: I don't think we need POST here. The form for the dates should use GET.

File: indico/modules/categories/controllers/display.py [hunk 8]
```diff
@@ -384,3 +386,24 @@ def _checkParams(self):
...
+
+class RHCategoryCalendarView(RHDisplayCategoryBase):
+    def _process_GET(self):
+        return WPCategory.render_template('display/calendar.html', self.category)
+
+    def _process_POST(self):
+        data = request.json
+        start = dateutil.parser.parse(data['start'])
+        end = dateutil.parser.parse(data['end'])
+        category_events = get_category_timetable([self.category.id], start, end, from_categ=self.category)['events']
```
Comment (ThiefMaster) [inline indico/modules/categories/controllers/display.py]: If we only care for events, maybe it'd be a good idea to add a kwarg `only_events` to that function? Getting sessions and contributions is rather expensive and if we don't need them there's no good reason to query them.

File: indico/modules/categories/blueprint.py (hunk 7 above)
Comment (mic4ael) [inline indico/modules/categories/blueprint.py]: forgot to remove it after we got rid of the form from that page

File: indico/htdocs/sass/modules/_category.scss [hunk 9]
```diff
@@ -579,3 +596,19 @@ ul.category-resource-qtip {
         width: 192px;
     }
 }
+
+.category-calendar-view {
+    .category-content {
+        max-width: 70%;
+
+        #event-calendar {
```
Comment (ThiefMaster) [inline indico/htdocs/sass/modules/_category.scss]: I'd use `.fc` here (it's added to that element by the calendar widget)

**Pull request merged**